### PR TITLE
Create requester approval flow

### PR DESCRIPTION
**Pull request by bzzbbz**
**tl;dr: This pull request establishes the requester approval flow.** 

### Summary
- Requesters can approve/reject tasks on the Manage Responses table
- Requesters can also review tasks, which takes them to a read-only Form page documenting individual tasker's responses. A video demo is provided below.

https://user-images.githubusercontent.com/46664399/165481152-19e1671c-eaf6-454d-96b9-9d14e3851c4d.mov

### Next Steps
- @christine-sun @jennsun You may need to expand the functions "handleApproveResponse" and "handleRejectResponse" in `pages/requester/manage-task/[taskid].tsx` to include Web3 Escrow services.
- @bzzbbz Add analytics / excel downloading features.

### CODE REVIEW

File: pages/requester/my-tasks.tsx [hunk 1]
```diff
@@ -110,7 +110,7 @@ export default function MyTasks() {
             size="small"
```
Comment (christine-sun) [inline pages/requester/my-tasks.tsx:110]: To improve some of the wording, on this page could the secondary button be "Close" and primary button be "Responses"

**Review by christine-sun — APPROVED** [review]

File: components/tables/ApprovalTableComponents.tsx [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+
+  if (props.data.length === 0)
+    return (
+      <Typography
+        textAlign="center"
+        color="primary"
+        fontWeight={400}
+        fontSize={20}
+      >
+        No Responses to display &#128532;
```
Comment (jennsun) [inline components/tables/ApprovalTableComponents.tsx:75]: what does "&#128532" represent?

File: src/Types.ts [hunk 3]
```diff
@@ -49,6 +50,8 @@ export type TaskStatus = 0 | 1 | 2 | 3;
 
 export type CreatedTaskStatus = 0 | 1 | 2;
 
+export type ApprovalStatus = 1 | 2 | 3;
```
Comment (jennsun) [inline src/Types.ts:53]: Can these start at 0, just like the CreatedTaskStatuses?

**Review by jennsun — CHANGES_REQUESTED** [review]
had 2 quick comments, otherwise this looks good!

File: cloud/cloud.js [hunk 4]
```diff
@@ -889,3 +892,203 @@ Moralis.Cloud.define(
...
+      return {
+        success: false,
+        message: "Approval update failed - unauthorized update!"
+      };
+    }
+    tableName = "TaskUsers";
+    const TaskUsers = Moralis.Object.extend(tableName);
+    const query = new Moralis.Query(TaskUsers);
+    query.get(request.params.objectId).then(
+      (taskUser) => {
```
Comment (nicholaspad) [inline cloud/cloud.js:1010]: Here, can you first check that `taskUser.get("status")` is `1` (i.e. not reviewed), and if it's not, then return an error? I think that this would ensure correct behavior: the status can only be changed from `1` to something else.

File: src/Types.ts (hunk 3 above)
Comment (nicholaspad) [reply src/Types.ts:53]: @bzzbbz correct me if I misunderstood, but I think that this is intentionally `1 | 2 | 3` so that the status states match up between requester and tasker:

Tasker: In progress (0), Pending verification (1), Verified & paid (2), Abandoned (3)

Requester: Not reviewed (1), Approved (2), Rejected (2)

Status of 0 isn't relevant for requesters.

File: components/tables/ApprovalTableComponents.tsx (hunk 2 above)
Comment (nicholaspad) [reply components/tables/ApprovalTableComponents.tsx:75]: It's an emoji

**Review by nicholaspad — CHANGES_REQUESTED** [review]
looks great! just some nits and questions :)

File: src/Types.ts (hunk 3 above)
Comment (bzzbbz) [reply src/Types.ts:53]: It's done intentionally to match the "status" column in the TaskUsers table in the database. Status of 0 in TaskUsers means that the user has started the task, but not completed it, so these tasks are filtered out of the approval table. Status 1 means that the task if completed, but no one had viewed it yet.

File: cloud/cloud.js [hunk 5]
```diff
@@ -1,3 +1,6 @@
+import { get } from "http";
+import Moralis from "moralis/types";
+
```
[the remark this answers is not in the document]
Comment (bzzbbz) [reply cloud/cloud.js:3]: Oh sorry I don't know why they're there

File: cloud/cloud.js [hunk 6]
```diff
@@ -889,3 +892,203 @@ Moralis.Cloud.define(
...
+  async (request) => {
+    const ethAddress = request.user.get("ethAddress");
+    return await checkRequesterCreatedTask(ethAddress, request.params.taskId)
+  },
+  {
+    fields: ["taskId"],
+    requireUser: true,
+  }
+);
+
```
[the remark this answers is not in the document]
Comment (bzzbbz) [reply cloud/cloud.js:921]: done

File: components/tables/ApprovalTableComponents.tsx [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+    },
+    {
+      field: "duration",
+      sortable: false,
+      disableColumnMenu: true,
+      type: "string",
+      minWidth: 150,
+      renderHeader: () => <TableHeader>Time Spent</TableHeader>,
+      renderCell: (params: GridValueGetterParams) => (
+        <TableCell>{params.row.duration} Minutes</TableCell>
```
[the remark this answers is not in the document]
Comment (bzzbbz) [reply components/tables/ApprovalTableComponents.tsx:46]: done

File: pages/requester/manage-task/[taskid].tsx [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+          {approvalStatusMap[params.row.status as ApprovalStatus]}
+        </TableCell>
+      ),
+    },
+    {
+      field: "",
+      headerName: "",
+      sortable: false,
+      disableColumnMenu: true,
+      minWidth: data?.some((e) => e.status === 1) ? 290 : 165,
```
[the remark this answers is not in the document]
Comment (bzzbbz) [reply pages/requester/manage-task/[taskid].tsx:101]: done

File: pages/requester/manage-task/[taskid].tsx [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+import { useEffect, useState } from "react";
+import { useRouter } from "next/router";
+import { useMoralis } from "react-moralis";
+import PrimaryButtonCTA from "../../../components/buttons/PrimaryButtonCTA";
+import SecondaryButtonCTA from "../../../components/buttons/SecondaryButtonCTA";
+import LoadingOverlay from "../../../components/common/LoadingOverlay";
+import PageHeader from "../../../components/common/PageHeader";
+import { TableCell, TableHeader } from "../../../components/tables/Helpers";
+import {
+  getTaskResponses,
```
[the remark this answers is not in the document]
Comment (bzzbbz) [reply pages/requester/manage-task/[taskid].tsx:12]: done

File: pages/requester/manage-task/[taskid]/[userId].tsx [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+  updateApprovalStatus,
+} from "../../../../src/Database";
+import LoadingOverlay from "../../../../components/common/LoadingOverlay";
+import {
+  GenericResponse,
+  QuestionType,
+  TaskProps,
+} from "../../../../src/Types";
+import { useFormik } from "formik";
+import handler from "../../../api/hello";
```
[the remark this answers is not in the document]
Comment (bzzbbz) [reply pages/requester/manage-task/[taskid]/[userId].tsx:22]: done

File: pages/requester/manage-task/[taskid]/[userId].tsx [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+
+export default function TaskerForm() {
+  const router = useRouter();
+  const { taskid, userId } = router.query;
+  const taskId = taskid;
+  const { isInitialized, Moralis, user } = useMoralis();
+  const [openLoading, setOpenLoading] = useState(false);
+  const [openAbandonLoading, setOpenAbandonLoading] = useState(false);
```
[the remark this answers is not in the document]
Comment (bzzbbz) [reply pages/requester/manage-task/[taskid]/[userId].tsx:30]: done

File: pages/requester/manage-task/[taskid]/[userId].tsx [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+  TaskProps,
+} from "../../../../src/Types";
+import { useFormik } from "formik";
+import handler from "../../../api/hello";
+
+export default function TaskerForm() {
+  const router = useRouter();
+  const { taskid, userId } = router.query;
+  const taskId = taskid;
+  const { isInitialized, Moralis, user } = useMoralis();
```
[the remark this answers is not in the document]
Comment (bzzbbz) [reply pages/requester/manage-task/[taskid]/[userId].tsx:28]: done

File: pages/requester/manage-task/[taskid]/[userId].tsx [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+  const [openLoading, setOpenLoading] = useState(false);
+  const [openAbandonLoading, setOpenAbandonLoading] = useState(false);
+  const [isAllowed, setIsAllowed] = useState(false);
+  const [data, setData] = useState<TaskProps>();
+
+  const handleApproveResponse = async (
```
[the remark this answers is not in the document]
Comment (bzzbbz) [reply pages/requester/manage-task/[taskid]/[userId].tsx:34]: done

File: pages/requester/manage-task/[taskid]/[userId].tsx [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+      setOpenLoading(false);
+      alert(res.message);
+      return;
+    }
+
+    alert(res.message);
+    setOpenLoading(false);
+  };
+
+  const handleRejectResponse = async (
```
[the remark this answers is not in the document]
Comment (bzzbbz) [reply pages/requester/manage-task/[taskid]/[userId].tsx:59]: done

File: pages/requester/manage-task/[taskid].tsx [hunk 15]
```diff
@@ -0,0 +1,213 @@
...
+        });
+      }
+      console.log(tempData);
+      setData(tempData);
+    });
+  }, [isInitialized, Moralis, refreshTable, isReady]);
+
+  return (
+    <>
+      <PageHeader title="Requester Created Tasks" />
```
[the remark this answers is not in the document]
Comment (bzzbbz) [reply pages/requester/manage-task/[taskid].tsx:190]: done

File: pages/requester/manage-task/[taskid]/[userId].tsx [hunk 16]
```diff
@@ -0,0 +1,123 @@
...
+      }
+    );
+  }, [isInitialized, Moralis, taskId]);
+
+  // if (!isAllowed) return <LoadingOverlay open={true} text="Verifying..." />;
+  if (!data) return <LoadingOverlay open={true} text="Loading Task..." />;
+
+  return (
+    <>
+      <PageHeader title="Task" />
```
[the remark this answers is not in the document]
Comment (bzzbbz) [reply pages/requester/manage-task/[taskid]/[userId].tsx:107]: done

File: pages/requester/manage-task/[taskid]/[userId].tsx [hunk 17]
```diff
@@ -0,0 +1,123 @@
...
+      return;
+    }
+
+    alert(`Successfully rejected response from user ${taskerId}!`);
+    setOpenLoading(false);
+  };
+
+  useEffect(() => {
+    if (!isInitialized || !taskId) return;
+    console.log(taskId);
```
[the remark this answers is not in the document]
Comment (bzzbbz) [reply pages/requester/manage-task/[taskid]/[userId].tsx:86]: done

File: pages/requester/manage-task/[taskid]/[userId].tsx [hunk 18]
```diff
@@ -0,0 +1,123 @@
...
+    <>
+      <PageHeader title="Task" />
+      <PrimaryButtonCTA
+        text="← Manage Responses"
+        size="small"
+        to={`/requester/manage-task/${taskId}`}
+        sx={{ mx: "auto", mt: 2 }}
+      />
+      <LoadingOverlay open={openLoading} text="Submitting Task..." />
+      <LoadingOverlay open={openAbandonLoading} text="Abandoning Task..." />
```
[the remark this answers is not in the document]
Comment (bzzbbz) [reply pages/requester/manage-task/[taskid]/[userId].tsx:115]: done

File: cloud/cloud.js (hunk 4 above)
Comment (bzzbbz) [reply cloud/cloud.js:1010]: done

File: cloud/cloud.js [hunk 19]
```diff
@@ -857,50 +857,205 @@ Moralis.Cloud.define(
...
+    const query = new Moralis.Query(TaskUsers);
+    return query.get(request.params.objectId).then(
+      (taskUser) => {
+        if (taskUser.get("status") !== 1) {
+          return {
+            success: false, 
+            message: "Cannot update status of tasks that are already approved/rejected"
+          };
+        }
+        taskUser.set("status", request.params.newStatus)
```
Comment (nicholaspad) [inline cloud/cloud.js:970]: Can you first check that `request.params.newStatus` is either `2` or `3`, and if not then return an error, e.g.

> Task status must be updated to either "approved" or "rejected"

File: cloud/cloud.js [hunk 20]
```diff
@@ -857,50 +857,205 @@ Moralis.Cloud.define(
...
+            localField: "questionId",
+            foreignField: "_id",
+            as: "question",
+          },
+        },
+      ];
+      return await query.aggregate(pipeline);
+    }
+    const responses = await aggregateResponses(taskId, taskerId);
+    console.log(responses);
```
Comment (nicholaspad) [inline cloud/cloud.js:1029]: Nit: remove log

File: pages/requester/manage-task/[taskid]/[userId].tsx [hunk 21]
```diff
@@ -0,0 +1,58 @@
+import { Container, Box } from "@mui/material";
```
Comment (nicholaspad) [inline pages/requester/manage-task/[taskid]/[userId].tsx:1]: Nit: remove unused `Box` import

File: pages/requester/manage-task/[taskid]/[userId].tsx [hunk 22]
```diff
@@ -0,0 +1,58 @@
...
+import Question from "../../../../components/taskerForm/Question";
+import PrimaryButtonCTA from "../../../../components/buttons/PrimaryButtonCTA";
+import { useEffect, useState } from "react";
+import PageHeader from "../../../../components/common/PageHeader";
+import { useRouter } from "next/router";
+import { useMoralis } from "react-moralis";
+import {
+  getTaskFormDataForView,
+  updateApprovalStatus,
```
Comment (nicholaspad) [inline pages/requester/manage-task/[taskid]/[userId].tsx:10]: Nit: remove unused `updateApprovalStatus` import

File: src/Database.ts [hunk 23]
```diff
@@ -1,4 +1,5 @@
 import MoralisType from "moralis";
+import TaskDetails from "../pages/tasker/task-details/[taskId]";
```
Comment (nicholaspad) [inline src/Database.ts:2]: Nit: remove unused `TaskDetails` import

**Review by nicholaspad — CHANGES_REQUESTED** [review]

**Review by jennsun — APPROVED** [review]

**Comment by jennsun** [comment]
comment regarding 
@christine-sun @jennsun You may need to expand the functions "handleApproveResponse" and "handleRejectResponse" in pages/requester/manage-task/[taskid].tsx to include Web3 Escrow services:

The approve/reject functions would not change the web3 component of the application, but we would have to implement additional checks before calling the smart contract functions

File: cloud/cloud.js (hunk 20 above)
Comment (bzzbbz) [reply cloud/cloud.js:1029]: done

File: pages/requester/manage-task/[taskid]/[userId].tsx (hunk 21 above)
Comment (bzzbbz) [reply pages/requester/manage-task/[taskid]/[userId].tsx:1]: done

File: src/Database.ts (hunk 23 above)
Comment (bzzbbz) [reply src/Database.ts:2]: done

**Review by nicholaspad — APPROVED** [review]
lgtm
